src/EventStore.Core/Services/Transport/Enumerators/Enumerator.StreamSubscription.cs

```diff
@@ -129,21 +132,120 @@ public class StreamSubscription<TStreamId> : StreamSubscription {
+					else
+						lastSubscriptionEventNumber = checkpointEventNumber;
+
+					while (true) {
+						lastSubscriptionEventNumber = await CatchUp(lastSubscriptionEventNumber);
+
+						if (!TrySyncLiveWithCatchup(lastCatchupEventNumber: lastSubscriptionEventNumber,
+							    ref lastLiveEventInfo))
+							continue;
+
+						(lastSubscriptionEventNumber, lastLiveEventInfo) = await GoLive(lastLiveEventInfo);
+					}
+				} catch (Exception ex) {
+					if (ex is not (OperationCanceledException or ReadResponseException.StreamDeleted))
+						Log.Error(ex, "An error has occurred in subscription to stream: {streamName}", _streamName);
+					Fail(ex);
+				} finally {
+					Log.Information("Subscription {subscriptionId} to {streamName} has ended.", _subscriptionId, _streamName);
 				}
 			}
 
-			private void CatchUp(StreamRevision? startRevision) {
-				Log.Information(
+			private async Task NotifyCaughtUp(long atEventNumber) {
+				Log.Debug(
+					"Live subscription {subscriptionId} to {streamName} caught up at {streamRevision} because the end of stream was reached.",
+					_subscriptionId, _streamName, atEventNumber);
+
+				await _channel.Writer.WriteAsync(new ReadResponse.SubscriptionCaughtUp(), _cancellationToken);
+			}
+
+			private async Task NotifyFellBehind(long atEventNumber) {
+				Log.Debug(
+					"Subscription {subscriptionId} to {streamName} fell behind at {streamRevision}.",
+					_subscriptionId, _streamName, atEventNumber);
+
+				await _channel.Writer.WriteAsync(new ReadResponse.SubscriptionFellBehind(), _cancellationToken);
+			}
+
+			private bool TrySyncLiveWithCatchup(long lastCatchupEventNumber, ref LiveEventInfo lastLiveEventInfo) {
+				while (_liveEvents.Reader.TryRead(out var liveEvent)) {
+					// drain all live events that are up to the last catch-up event
+					// (or the next one if the catch-up event is not in the live channel)
+					// while keeping track of the last drained event.
+					lastLiveEventInfo.EventNumber = liveEvent.resolvedEvent.OriginalEventNumber;
+					lastLiveEventInfo.SequenceNumber = liveEvent.sequenceNumber;
+					if (liveEvent.resolvedEvent.OriginalEventNumber >= lastCatchupEventNumber)
+						break;
+				}
+
+				// we can switch to live if the last live event is the last catch-up event
+				return lastLiveEventInfo.EventNumber == lastCatchupEventNumber;
+			}
+
+			private async ValueTask<(long lastSubscriptionEventNumber, LiveEventInfo lastLiveEventInfo)> GoLive(LiveEventInfo lastLiveEventInfo) {
+				var lastSubscriptionEventNumber = lastLiveEventInfo.EventNumber;
+				await NotifyCaughtUp(lastSubscriptionEventNumber);
+
+				await foreach (var liveEvent in _liveEvents.Reader.ReadAllAsync(_cancellationToken)) {
+					bool sequenceCorrect = liveEvent.sequenceNumber == lastLiveEventInfo.SequenceNumber + 1;
+					lastLiveEventInfo = new LiveEventInfo(liveEvent.resolvedEvent.OriginalEventNumber, liveEvent.sequenceNumber);
```

Comment:
prefer to avoid the allocation on each live event here. it might be sensible to make LiveEventInfo a `record struct` and remove  the explicit get/set (and just construct a new one in TrySyncLiveWithCatchup)

Reply:
fixed :+1: turns out the compiler doesn't complain about passing a `ref record struct` as parameter to `TrySyncLiveWithCatchup`, so i've not made any changes there. i think that's because the lastLiveEventInfo is on the stack and thus the stack address of the struct can be passed in.